File: apps/api/services/highlight_matches.py
"""Audit search highlighter (cycle WW1, Python half).

Server-side mirror of `apps/web/lib/highlight-matches.ts`. Used
by the Slack alert digest excerpt builder, the email digest's
search-result preview, and the audit row plaintext export's
highlight markers.

  find_matches(text, terms)  — tuple[Match, ...]
  Match                      — frozen dataclass: (start, end)

Composes with II1's `parse_search_query.free_text`.

Pure stdlib.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Match:
    """A non-overlapping match span in the original text.

    `start` is inclusive, `end` is exclusive. Indexes refer to
    the ORIGINAL `text` (NOT the lowercased version) — caller
    can slice directly.
    """

    start: int
    end: int


def find_matches(
    text: str | None,
    terms: list[str],
) -> tuple[Match, ...]:
    """Find non-overlapping match spans for any of `terms` in `text`.

    Algorithm:
      1. Lowercase `text` and each `term` for case-insensitive
         matching.
      2. Collect ALL candidate matches (overlapping allowed).
      3. Sort by (start ASC, length DESC) so earlier wins; on
         tie, longer term wins.
      4. Greedy left-to-right select non-overlapping spans.

    Returns matches in document order.
    """
    if not text or not terms:
        return ()

    text_lower = text.lower()
    candidates: list[tuple[int, int]] = []

    for term in terms:
        if not term:
            continue
        t = term.strip().lower()
        if not t:
            continue
        idx = 0
        while idx < len(text_lower):
            found = text_lower.find(t, idx)
            if found < 0:
                break
            candidates.append((found, found + len(t)))
            idx = found + 1

    if not candidates:
        return ()

    # Sort: start ASC, length DESC (longer wins on same start).
    candidates.sort(key=lambda c: (c[0], -(c[1] - c[0])))

    result: list[Match] = []
    last_end = 0
    for start, end in candidates:
        if start >= last_end:
            result.append(Match(start=start, end=end))
            last_end = end

    return tuple(result)

File: apps/web/lib/highlight-matches.ts
/**
 * Audit search highlighter (cycle WW1, TS half).
 *
 * Given free-text search terms + a text body, return non-
 * overlapping spans of matches for UI highlighting in the
 * audit list note column, the search-results page, and the
 * Slack alert digest excerpt builder.
 *
 *   findMatches(text, terms)  — readonly Match[]
 *   Match                     — { start, end }
 *
 * Composes with II1's `parse_search_query.free_text` (caller
 * passes `query.free_text` as `terms`).
 *
 * Pure TS. Mirrors `apps/api/services/highlight_matches.py`.
 *
 * Pinned invariants:
 *   * Case-insensitive matching.
 *   * Non-overlapping spans (greedy left-to-right; on tie,
 *     longer wins).
 *   * Empty / null `text` or empty `terms` → [].
 *   * Whitespace-only term skipped.
 *   * Deterministic order: matches in document order.
 *   * Cross-language byte-for-byte parity.
 */


export interface Match {
  /** Inclusive start index. */
  start: number;
  /** Exclusive end index. */
  end: number;
}


/**
 * Find non-overlapping match spans for any of `terms` in `text`.
 *
 * Returns a list of `{start, end}` matches in document order.
 * Indexes refer to the ORIGINAL `text` (NOT the lowercased
 * version) — caller can slice directly.
 *
 *   * findMatches("Hello World", ["hello"])      → [{start: 0, end: 5}]
 *   * findMatches("foo bar foo", ["foo"])        → [{0,3}, {8,11}]
 *   * findMatches("aaaa", ["aa"])                → [{0,2}, {2,4}]   (greedy)
 *   * findMatches("hello", ["hello", "ell"])     → [{0,5}]          (longer wins)
 *   * findMatches(null, [...])                   → []
 *   * findMatches("text", [])                    → []
 */
export function findMatches(
  text: string | null | undefined,
  terms: readonly string[],
): readonly Match[] {
  if (!text || terms.length === 0) return [];

  const textLower = text.toLowerCase();
  const candidates: Array<[number, number]> = [];

  for (const term of terms) {
    if (!term) continue;
    const t = term.trim().toLowerCase();
    if (!t) continue;
    let idx = 0;
    while (idx < textLower.length) {
      const found = textLower.indexOf(t, idx);
      if (found < 0) break;
      candidates.push([found, found + t.length]);
      // Step forward by 1 to find overlapping candidates;
      // greedy resolution below picks non-overlapping.
      idx = found + 1;
    }
  }

  if (candidates.length === 0) return [];

  // Sort: start ASC, length DESC (longer wins on tie at same start).
  candidates.sort((a, b) => {
    if (a[0] !== b[0]) return a[0] - b[0];
    return b[1] - b[0] - (a[1] - a[0]);
  });

  // Greedy non-overlapping selection.
  const result: Match[] = [];
  let lastEnd = 0;
  for (const [start, end] of candidates) {
    if (start >= lastEnd) {
      result.push({ start, end });
      lastEnd = end;
    }
  }

  return result;
}
